cpp/src/arrow/testing/random.cc

```diff
@@ -343,8 +343,119 @@ struct DecimalGenerator {
   }
 };
 
+template <>
+struct DecimalGenerator<Decimal32Type> {
+  using DecimalBuilderType = typename TypeTraits<Decimal32Type>::BuilderType;
+  using DecimalValue = typename DecimalBuilderType::ValueType;
+
+  std::shared_ptr<DataType> type_;
+  RandomArrayGenerator* rng_;
+
+  static int32_t MaxDecimalInteger(int32_t digits) {
+    return static_cast<int32_t>(std::ceil(std::pow(10.0, digits))) - 1;
+  }
+
+  std::shared_ptr<Array> MakeRandomArray(int64_t size, double null_probability,
+                                         int64_t alignment, MemoryPool* memory_pool) {
+    static constexpr int32_t kMaxDigitsInInteger = 9;
+    static constexpr int kNumIntegers = Decimal32Type::kByteWidth / 4;
+    static_assert(
+        kNumIntegers == (Decimal32Type::kMaxPrecision + kMaxDigitsInInteger - 1) /
+                            (kMaxDigitsInInteger + 1),
+        "inconsistent decimal metadata: kMaxPrecision doesn't match kByteWidth");
+
+    // First generate separate random values for individual components:
+    // boolean sign (including null-ness), and uint64 "digits" in big endian order.
+    const auto& decimal_type = checked_cast<const DecimalType&>(*type_);
+
+    auto remaining_digits = decimal_type.precision();
+    auto values = checked_pointer_cast<Int32Array>(rng_->Int32(
+        size, -1 * MaxDecimalInteger(remaining_digits),
+        MaxDecimalInteger(remaining_digits), null_probability, alignment, memory_pool));
+
+    DecimalBuilderType builder(type_, memory_pool, alignment);
+    ABORT_NOT_OK(builder.Reserve(size));
+
+    for (int64_t i = 0; i < size; ++i) {
+      if (values->IsValid(i)) {
+        builder.UnsafeAppend(DecimalValue{values->Value(i)});
+      } else {
+        builder.UnsafeAppendNull();
+      }
+    }
+    std::shared_ptr<Array> array;
+    ABORT_NOT_OK(builder.Finish(&array));
+    return array;
+  }
+};
+
+template <>
+struct DecimalGenerator<Decimal64Type> {
```

Comment:
Hmm, can you perhaps refactor those two specialisations for `Decimal32Type` and `Decimal64Type` into a single one (perhaps `SmallDecimalGenerator<T>`)?

Reply:
I was able to create a single generic `SmallDecimalGenerator<T>` so this worked out :)